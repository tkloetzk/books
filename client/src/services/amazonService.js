import apiConfig from '../config/apiConfig';
import axios from 'axios';

export function getAmazonBookService(isbns) {
  return axios
    .post(apiConfig.amazon, { isbns })
    .then(books => books.data)
    .catch(error => {
      throw error;
    });
}

export function getAmazonBookServiceSingle(isbn) {
  return axios
    .post(apiConfig.amazonV2, { isbn })
    .then(res => res.data)
    .catch(error => {
      throw error;
    });
}
