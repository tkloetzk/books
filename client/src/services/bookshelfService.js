import apiConfig from '../config/apiConfig';
import axios from 'axios';
import sortBooklist from '../util/calculator';

export function getBookshelfService(excludedGenres) {
  return axios
    .post(apiConfig.bookshelf, excludedGenres)
    .then(res => sortBooklist(res.data))
    .catch(error => {
      throw error;
    });
}
export function addBookshelfService(booklist) {
  return axios
    .post(`${apiConfig.bookshelf}/add`, booklist)
    .then(res => console.log(res.data))
    .catch(error => {
      throw error;
    });
}

export function updateBookOnBookshelfService(id, fields) {
  return axios
    .put(`${apiConfig.bookshelf}/update/${id}`, fields)
    .then(res => console.log(res.data))
    .catch(error => {
      throw error;
    });
}
