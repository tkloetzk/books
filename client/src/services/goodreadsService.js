import apiConfig from '../config/apiConfig';
import axios from 'axios';

export function getGoodreadsSingleBooksService(booklist) {
  return axios
    .get(`${apiConfig.goodreads}/${booklist}`)
    .then(booklist => booklist.data)
    .catch(error => {
      throw error;
    });
}
