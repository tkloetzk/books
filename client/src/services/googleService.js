import apiConfig from '../config/apiConfig';
import axios from 'axios';

export function getGoogleBookService(isbn) {
  return axios
    .get(`${apiConfig.google}/${isbn}`)
    .then(book => book.data)
    .catch(error => {
      throw error;
    });
}
