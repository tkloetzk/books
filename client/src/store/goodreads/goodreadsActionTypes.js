export const FETCH_GOODREADS_BOOK_HAS_ERRORED =
  'FETCH_GOODREADS_BOOK_HAS_ERRORED';
export const FETCH_GOODREADS_BOOK_IS_LOADING =
  'FETCH_GOODREADS_BOOK_IS_LOADING';
export const FETCH_GOODREADS_BOOKS_SUCCESS = 'FETCH_GOODREADS_BOOKS_SUCCESS';
export const FETCH_GOODREADS_BOOK_SUCCESS = 'FETCH_GOODREADS_BOOK_SUCCESS';
export const CLEAR_GOODREADS_BOOKS_SUCCESS = 'CLEAR_GOODREADS_BOOKS_SUCCESS';
