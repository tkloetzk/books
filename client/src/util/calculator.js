import get from 'lodash/get';
import forEach from 'lodash/forEach'
import clone from 'lodash/clone'

let meanGoodreadsVotes
let meanAmazonVotes
let totalMean

export default function sortBooklist(originalBooklist) {
  const booklist = clone(originalBooklist);
  calculateMeans(booklist);
  calculateAdjustedRating(booklist);
  booklist.sort((a, b) =>
    a.adjustedRating < b.adjustedRating
      ? 1
      : b.adjustedRating < a.adjustedRating
      ? -1
      : 0
  );
  return booklist;
}

function calculateMeans(booklist) {
  let goodreadsVotes = 0,
    amazonVotes = 0,
    total = 0;

  forEach(booklist, book => {
    goodreadsVotes += get(book, 'goodreadsAverageRating')
    amazonVotes += get(book, 'amazonAverageRating');
    total +=
      get(book, 'goodreadsAverageRating') +
      get(book, 'amazonAverageRating');
  })

  meanGoodreadsVotes = goodreadsVotes / booklist.length;
  meanAmazonVotes = amazonVotes / booklist.length;
  totalMean = total / 2 / booklist.length;

  return booklist;
}

function calculateAdjustedRating(booklist) {
  for (var i = 0; i < booklist.length; i++) {
    var book = booklist[i];
    var adjustedRating =
      (getAdjustedRating(
        get(book, 'goodreadsRatingsCount'),
        get(book, 'goodreadsAverageRating'),
        meanGoodreadsVotes,
        trimmean(getGoodreadsRatingsCountList(booklist))
      ) +
        getAdjustedRating(
          get(book, 'amazonRatingsCount'),
          get(book, 'amazonAverageRating'),
          meanAmazonVotes,
          trimmean(getAmazonRatingsCountList(booklist))
        )) /
      2;
    book.adjustedRating = adjustedRating;
  }
}

function getAdjustedRating(ratingsCount, averageRating, meanVote, minVotes) {
  //double rating = averageRating * 2;
  return (
    (ratingsCount / (ratingsCount + minVotes)) * averageRating +
    (minVotes / (ratingsCount + minVotes)) * meanVote
  );
}

function getGoodreadsRatingsCountList(booklist) {
  let votes = [];
  for (const book of booklist) {
    votes.push(book.goodreadsRatingsCount);
  }

  return votes.sort(function(a, b) {
    return a - b;
  });
}

function getAmazonRatingsCountList(booklist) {
  let votes = [];
  for (const book of booklist) {
    votes.push(book.amazonRatingsCount);
  }

  return votes.sort(function(a, b) {
    return a - b;
  });
}

function trimmean(votes) {
  votes.sort(function(a, b) {
    return a - b;
  });

  const removeAmount = (votes.length * 0.25) / 2;
  const roundDownNearedMultipleTwo = Math.floor(
    removeAmount >= 0
      ? (removeAmount / 2) * 2
      : ((removeAmount - 2 + 1) / 2) * 2
  );
  return getMean(
    votes.slice(
      roundDownNearedMultipleTwo,
      votes.length - roundDownNearedMultipleTwo
    )
  );
}

function getMean(array) {
  let sum = 0;
  for (let i = 0; i < array.length; i++) {
    sum += array[i];
  }
  return sum / array.length;
}
