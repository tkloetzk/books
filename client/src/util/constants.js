export const LOADING_STATUSES = {
  initial: 'initial',
  loading: 'loading',
  success: 'success',
  errored: 'errored',
};
