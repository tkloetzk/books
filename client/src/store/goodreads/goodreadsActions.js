import * as types from './goodreadsActionTypes';
import {
  getGoodreadsBooksService,
  getGoodreadsSingleBooksService,
} from '../../services/goodreadsService';
import sortBooklist from '../bookshelf/bookshelfActions';

export function getGoodreadsBookFailure(bool) {
  return {
    type: types.FETCH_GOODREADS_BOOK_HAS_ERRORED,
    hasErrored: bool,
  };
}

export function getGoodreadsBookIsLoading(bool) {
  return {
    type: types.FETCH_GOODREADS_BOOK_IS_LOADING,
    isLoading: bool,
  };
}

export function getGoodreadsBooksSuccess(booklist) {
  return {
    type: types.FETCH_GOODREADS_BOOKS_SUCCESS,
    booklist,
  };
}

export function getGoodreadsBooks(booklist) {
  return dispatch => {
    dispatch(getGoodreadsBookIsLoading(true));
    return getGoodreadsBooksService(booklist)
      .then(resp => {
        dispatch(getGoodreadsBookIsLoading(false));
        const sortedResp = sortBooklist(resp);
        console.log('sorted', sortedResp);
        // dispatch(getGoodreadsBooksSuccess(resp));
      })
      .catch(error => {
        dispatch(getGoodreadsBookIsLoading(false));
        dispatch(getGoodreadsBookFailure(true));
      });
  };
}

export function getGoodreadsBookSuccess(book) {
  return {
    type: types.FETCH_GOODREADS_BOOK_SUCCESS,
    book,
  };
}

export function getGoodreadsBook(isbn) {
  return dispatch => {
    dispatch(getGoodreadsBookIsLoading(true));
    return getGoodreadsSingleBooksService(isbn)
      .then(resp => {
        dispatch(getGoodreadsBookIsLoading(false));
        dispatch(getGoodreadsBookSuccess(resp));
      })
      .catch(error => {
        dispatch(getGoodreadsBookIsLoading(false));
        dispatch(getGoodreadsBookFailure(true));
      });
  };
}
