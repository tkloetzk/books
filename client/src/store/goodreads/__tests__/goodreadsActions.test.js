import * as actions from '../goodreadsActions';
import * as types from '../goodreadsActionTypes';
import { LOADING_STATUSES } from '../../../util/constants';
import { getGoodreadsSingleBooksService } from '../../../services/goodreadsService';
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'

const createMockStore = configureMockStore([thunk])

jest.mock('../../../services/goodreadsService')

describe('goodreads actions', () => {
  let book;
  beforeEach(() => {
    book = {
      isbn: '9781400079094',
      goodreadsAverageRating: 3.5,
      goodreadsRatingsCount: 90,
    };
    getGoodreadsSingleBooksService.mockClear()
  });
  it('can set goodreads book hasErrored', () => {
    
  })
})
