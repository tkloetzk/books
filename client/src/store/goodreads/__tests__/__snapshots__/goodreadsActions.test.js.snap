// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`goodreads actions can clear goodreads books 1`] = `
Array [
  Array [
    Object {
      "type": "CLEAR_GOODREADS_BOOKS_SUCCESS",
    },
  ],
]
`;
